File: FruitIntoBaskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int sz = fruits.size();
        int maxFruit = 0;
        int fruit1 = fruits[0], fruit1index = 0;
        int fruit2 = -1, fruit2index = -1;

        int i = 1;
        for(;i < sz;i++){
            if(fruits[i] != fruits[i - 1]){
                fruit2 = fruits[i];
                fruit2index = i;
                break;
            }
            else
                fruit1index++;
        }

        int length = i;
        if(length >= sz)
            return length;
        
        for(;i < sz;i++){
            if(fruits[i] == fruit1){
                fruit1index = i;
            }
            else if(fruits[i] == fruit2){
                fruit2index = i;
            }
            else if(fruits[i - 1] == fruit1){
                maxFruit = max(maxFruit, length);
                length = fruit1index - fruit2index;
                fruit2 = fruits[i];
                fruit2index = i;
            }
            else /*if(fruits[i - 1] == fruit2)*/{
                maxFruit = max(length, maxFruit);
                length = fruit2index - fruit1index;
                fruit1 = fruits[i];
                fruit1index = i;
            }
            
            length++;
        }
        maxFruit = max(length, maxFruit);
        return maxFruit;
    }
};

File: phase_one_eduacation/week4/leetcode/fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        res = 0
        unique = 0
        window = defaultdict(int)

        left = 0
        for right in range(len(fruits)):
            if fruits[right] not in window:
                unique += 1
            window[fruits[right]] += 1

            while unique > 2:
                window[fruits[left]] -= 1
                if window[fruits[left]] == 0:
                    del window[fruits[left]]
                    unique -= 1
                    
                left += 1
            
            res = max(res, right - left + 1)
        
        return res